neutone_sdk/filters.py

```diff
@@ -0,0 +1,266 @@
+        sample_rate: int,
+    ):
+        """Feed into time-varying svf
+
+        Args:
+            audio (torch.Tensor): Input audio [batch_size (or n_channels), n_samples]
+            cutoff (torch.Tensor): Cutoff frequency [batch_size, n_samples, 1]
+            resonance (torch.Tensor): Resonance (0 ~ 1), [batch_size, n_samples, 1]
+            mix (torch.Tensor): Mix coeff. bp, lp and hp [batch_size, n_samples, 3] ex.) [[[1.0, 0.0, 0.0]]] = bandpass
+
+        Returns:
+            audio (torch.Tensor): [n_channels, n_samples]
+        """
+        cutoff = torch.clamp(cutoff, min=1e-4, max=sample_rate / 2 - 1e-4)
+        resonance = torch.clamp(resonance, min=1e-4)
+        g = torch.tan(math.pi / sample_rate * cutoff).permute(1, 0, 2)
+        twoR = 1 / resonance.permute(1, 0, 2)
+        mix = mix.permute(1, 0, 2)
+        return self.svf(audio.permute(1, 0), g, twoR, mix)
+
+
+class _SVFLayer(nn.Module):
+    """
+    SVF implementation based on "Time-varying filters for musical applications" [Wishnick, 2014]
+    NOTE: This SVF is slow for use in training due to recurrent operations
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.register_buffer("state", torch.zeros(1, 2))
+
+    def forward(
+        self,
+        audio: torch.Tensor,
+        g: torch.Tensor,
+        twoR: torch.Tensor,
+        mix: torch.Tensor,
+    ):
+        """pass audio through SVF
+        Args:
+            *** time-first, batch-second ***
+            audio (torch.Tensor): [n_samples, batch_size]
+            All filter parameters are [n_samples, batch_size, 1 or 3 (mix)]
+            g (torch.Tensor): Normalized cutoff parameter
+            twoR (torch.Tensor): Damping parameter
+            mix (torch.Tensor): Mixing coefficient of bp, lp and hp
+
+        Returns:
+            [torch.Tensor]: Filtered audio. Shape [batch, n_samples]
+        """
+        seq_len, batch_size = audio.shape
+        T = 1.0 / (1.0 + g * (g + twoR))
+        H = T.unsqueeze(-1) * torch.cat(
+            [torch.ones_like(g), -g, g, twoR * g + 1], dim=-1
+        ).reshape(seq_len, batch_size, 2, 2)
+
+        # Y = gHBx + Hs
+        gHB = g * T * torch.cat([torch.ones_like(g), g], dim=-1)
+        # [n_samples, batch_size, 2]
+        gHBx = gHB * audio.unsqueeze(-1)
+
+        Y = torch.empty(seq_len, batch_size, 2, device=audio.device)
+        # initialize filter state
+        state = self.state.expand(batch_size, -1)
+        for t in range(seq_len):
+            Y[t] = gHBx[t] + torch.bmm(H[t], state.unsqueeze(-1)).squeeze(-1)
+            state = 2 * Y[t] - state
+        self.state = state
+
+        # HP = x - 2R*BP - LP
+        y_hps = audio - twoR.squeeze(-1) * Y[:, :, 0] - Y[:, :, 1]
+
+        y_mixed = (
+            twoR.squeeze(-1) * mix[:, :, 0] * Y[:, :, 0]
+            + mix[:, :, 1] * Y[:, :, 1]
+            + mix[:, :, 2] * y_hps
+        )
+        y_mixed = y_mixed.permute(1, 0).contiguous()
```

Comment:
Do we need this contiguous call?

Reply:
I put it in bc I needed it for some other projects, but it's unnecessary for this use case.